### ISSUE TITLE
binary-compat: dlang: Unhandled Trap 0 error

### ISSUE THREAD

**Issue by StefanJum** [unikraft/unikraft#495]
### Describe the bug

When running dlang static PIE binaries under binary compatibility mode, using [run-app-elfloader](https://github.com/unikraft/run-app-elfloader), it crashes with this message:
```
[    0.561820] CRIT: [libkvmplat] <trace.c @   41> RIP: 000000003f8701ae CS: 0008
[    0.563785] CRIT: [libkvmplat] <trace.c @   42> RSP: 000000003ffafb80 SS: 0010 EFLAGS: 00010293
[    0.566175] CRIT: [libkvmplat] <trace.c @   44> RAX: 000000003b9aca00 RBX: 000000003fa54f88 RCX: 0000000000000000
[    0.568947] CRIT: [libkvmplat] <trace.c @   46> RDX: 0000000000000000 RSI: 000000003ffafb80 RDI: 0000000000000001
[    0.571702] CRIT: [libkvmplat] <trace.c @   48> RBP: 0000000000000003 R08: 000000003f8700f0 R09: 0000000040020001
[    0.574433] CRIT: [libkvmplat] <trace.c @   50> R10: 000000003fa52f20 R11: 0000000000000246 R12: 000000003fc0dea0
[    0.577181] CRIT: [libkvmplat] <trace.c @   52> R13: 000000003fc017c0 R14: 0000000000000000 R15: 0000000000000001
[    0.580019] CRIT: [libkvmplat] <traps.c @   80> 3f8701ae <+0>: idiv %rcx
[    0.581823] CRIT: [libkvmplat] <traps.c @   80> 3f8701b1 <+3>: mov %rax, (%rbx)
[    0.583776] CRIT: [libkvmplat] <traps.c @   80> 3f8701b4 <+6>: test %rax, %rax
[    0.585699] CRIT: [libkvmplat] <traps.c @   80> 3f8701b7 <+9>: jnz 0x000000003F870144
[    0.587851] CRIT: [libkvmplat] <traps.c @   80> 3f8701b9 <+11>: add $0x20, %rsp
[    0.589852] CRIT: [libkvmplat] <traps.c @   80> 3f8701bd <+15>: pop %rbx
[    0.591690] CRIT: [libkvmplat] <traps.c @   80> 3f8701be <+16>: ret
[    0.593343] CRIT: [libkvmplat] <traps.c @   80> 3f8701bf <+17>: nop
[    0.595023] CRIT: [libkvmplat] <traps.c @   81> Crashing
```

The problem is caused by a division by 0, caused on [this line in the druntime environment](https://github.com/dlang/druntime/blob/master/old/src/core/time.d#L2852).
The problem occurs because the `clock_getres` syscall is not implemented in Unikraft, and it does not change the value of the `struct timespec *res` parameter, so when a division by `ts.tv_nsec` is made, the value of `ts.tv_nsec` is 0 (see issue #496).

This is solved by a hardcoded value (e.g. 1ns) for the `struct timespec *res` parameter in the [current version of the syscall](https://github.com/unikraft/unikraft/blob/staging/lib/uktime/time.c#L142), but it's just a hack, it does not solve the actual issue.

### Steps to reproduce

Use [run-app-elfloader](https://github.com/unikraft/run-app-elfloader) to run the dlang helloworld app. Use the command below, assuming the [run-app-elfloader repository](https://github.com/unikraft/run-app-elfloader) and the [static-pie-apps repository](https://github.com/unikraft/static-pie-apps/) are cloned in the same directory:
```bash
$ ./run_elfloader ../static-pie-apps/lang/rust/helloworld-gnu
```

For more information about building your own setup, see [this link](https://unikraft.org/community/hackathons/2022-05-lyon/bincompat/#00-setup).

### Expected behavior

The dlang helloworld application should run and print "Hello world!".

### Which architectures were you using or does this bug affect?

x86_64

### Which platforms were you using or does this bug affect?

kvm

**Comment by razvand** [comment]
@StefanJum , should this be closed, since #503 has been merged?

**Comment by StefanJum** [comment]
@razvand yes, this was solved by #503
